### PR TITLE
Make Asset generic

### PR DESCRIPTION
**Pull request by bananaturtlesandwich**
by leaving the reader generic, you can use a variety of types such as File or BufReader which means you don't have to load the entire file into memory and risk a vector buffer overflow

### CODE REVIEW

File: unreal_asset/src/reader/asset_trait.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ pub trait AssetTrait {
     fn get_custom_version<T>(&self) -> CustomVersion
     where
         T: CustomVersionTrait + Into<i32>;
-    fn position(&self) -> u64;
+    fn position(&mut self) -> u64;
```
Comment (localcc) [inline unreal_asset/src/reader/asset_trait.rs:16]: Fetching position should never mutate state

**Review by localcc — CHANGES_REQUESTED** [review]

**Comment by bananaturtlesandwich** [comment]
need to sort out the unit tests

File: unreal_asset/src/reader/asset_trait.rs (hunk 1 above)
Comment (bananaturtlesandwich) [reply unreal_asset/src/reader/asset_trait.rs:16]: it doesn't change position at all though - it just calls seek(SeekFrom::Current(0))

Comment (localcc) [reply unreal_asset/src/reader/asset_trait.rs:16]: it still shouldn't force you to pass a mutable reference, so another solution must be found

Comment (bananaturtlesandwich) [reply unreal_asset/src/reader/asset_trait.rs:16]: but it's only used internally so why should it matter?

Comment (localcc) [reply unreal_asset/src/reader/asset_trait.rs:16]: even internal apis should be correct and shouldn't require mutability on methods that do not mutate state

**Pull request closed without merging**